### PR TITLE
[W10-3] EventOrganiser

### PR DESCRIPTION
**Pull request by kengjichow**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddEventCommand.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Adds a person to the address book.
+ */
+public class AddEventCommand extends Command {
+
+    public static final String COMMAND_WORD = "addEvent";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD;
+    public static final String MESSAGE_SUCCESS = "New event added: %1$s";
+    public static final String MESSAGE_DUPLICATE_PERSON = "This event already exists in the address book";
```
Comment (AceCentury) [inline src/main/java/seedu/address/logic/commands/AddEventCommand.java]: Perhaps this variable should be named MESSAGE_DUPLICATE_EVENT ?

File: src/main/java/seedu/address/model/event/UniqueEventList.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.person.Person;
+import seedu.address.model.person.exceptions.DuplicatePersonException;
+import seedu.address.model.person.exceptions.PersonNotFoundException;
+
+/**
+ * A list of persons that enforces uniqueness between its elements and does not allow nulls.
```
Comment (AceCentury) [inline src/main/java/seedu/address/model/event/UniqueEventList.java]: Similarly, your javadoc comments should be updated to reflect that the function is for "Event"

File: src/test/java/seedu/address/logic/commands/AddCommandTest.java [hunk 3]
```diff
@@ -152,6 +163,11 @@ public void redoAddressBook() {
         public void commitAddressBook() {
             throw new AssertionError("This method should not be called.");
         }
+
+        @Override
+        public void addEvent(Event toAdd) {
```
Comment (AceCentury) [inline src/test/java/seedu/address/logic/commands/AddCommandTest.java]: You can consider creating a separate test class, e.g. AddEventCommandTest, to test the AddEventCommand class which you created. 

This would be more ideal as it would allow you to isolate issues quicker. E.g. if the AddCommandTest fails, you would need to identify whether it is the AddCommand **or** the AddEventCommand that is the one failing.

**Review by AceCentury — COMMENTED** [review]
Good progress so far, especially with the extensive code modifications!

Here are some things to take note:

* Try to replace the build badges (Travis, Appveyor etc) to with that of your team repo instead of AB4 repo.

* Your homepage (https://cs2103-ay1819s1-w10-3.github.io/main/) and netlify preview don't seem to be synchronised with your master branch. Check that Travis is set up correctly to deploy pages.

* Try to add a UI mockup to replace the existing one if possible.

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -831,29 +875,282 @@ Priorities: High (must have) - `* * \*`, Medium (nice to have) - `* \*`, Low (un
...
+
+*Actor: Event organiser, another user that will be invited to join the event organiser’s event*
+
+*Precondition: All the relevant users have identified themselves*
+
+*MSS:*
+
+1.	Event organiser searches for other users.
+2.	System shows user all the other users.
+3.	User chooses a user that he/she wants to invite
```
Comment (AceCentury) [inline docs/DeveloperGuide.adoc]: Try to have consistent punctuation. Some of your sentences end with full stops, some don't. I would suggest adding full stops all of sentences in all your use cases.

**Comment by AceCentury** [comment]
@yao-feng Don't forget to update your profile picture.

**Comment by yaofeng-wang** [comment]
Thanks for the reminder :)

**Comment by AceCentury** [comment]
I've noticed that you all had a number of commits to "Fix Style". 

I would advise you to run the tests locally first, fix any relevant errors **then** make the commit.

**Comment by adjscent** [comment]
> I've noticed that you all had a number of commits to "Fix Style".
> 
> I would advise you to run the tests locally first, fix any relevant errors **then** make the commit.

The issues do not appear in the local tests, and appear only on Travis CI.

**Comment by AceCentury** [comment]
Good job on your progress towards v1.1!

Here are some things to take note/update:

- Travis build status badge should point to the build status of your own repo, not se-edu/AB-4. Other badges should be updated to match your project or removed.

- The navigation bar at the top of your website preview should be tweaked/removed to reflect the homepage of your product.
image

- About Us page - @theJrLinguist your photo doesn't seem to be rendering properly on the netlify preview. @yao-feng if possible, try to update your photo such that your face is clearly visible. I find that your face is a little dim in the current photo.

- Developer Guide

All sentences in use cases should have full stops.
image

Non-functional requirements could be expanded a little bit.

Division of work should also have a consistent format. Some users don't have any designated "feature" (?) to work on.

image

**Comment by AceCentury** [comment]
Minor typo for the **Extensions** of U07

image

**Comment by AceCentury** [comment]
### v1.2 Feedback
The comments below are related to your Developer Guide

**Section 2: Design**
- I can see that some of the diagrams have been updated, but do remember to update all the outdated diagrams (e.g. UI)

**Section 3: Implementation**
- Would be good if you can update the Undo/Redo command references from AddressBook to EventOrganiser.

- For a sequence diagram in the **Poll Feature**, it would be good if the arrow could point to the middle of the object.

image

- For the sequence diagram in the **TimeTable Feature**, it would be good if the self-invocations have a return arrow.

image

- That same sequence diagram also happens to have an instance of AddressBookParser. I am not sure if it's intentional or it was overlooked, but just pointing it out. 

Overall, I can see that effort has been put in to describe the Implementations of specific components/features. It would be good if each component/feature could have their own class and sequence diagrams. 

So far, the **Poll Feature** is very well documented in terms of implementation and design considerations. Good job!

**Contact Us Page**

- Please update the details on the **Contact Us** page and link to your issue tracker.

image

**Comment by AceCentury** [comment]
**Following comments are in relation to Developer Guide**

**Chapter 2: Design**

- I did not observe any changes in this section except for **Model and Storage**. Correct me if I may have missed anything.
- For the diagram in **Model Component**, the multiplicity is somewhat unclear. Are you implying that both have a multiplicity of 1? Ideally, try to be more explicit and put "1" at both ends if that is the case.

image


**3.1. Login/logout feature**

- The sequence diagram is very small. Yes, I can always zoom in, but not sure if you might be penalised for that.

- For self-invocations in your sequence diagram, I am unsure if showing it as "stacked bars" instead of "half overlapping" bars is accepted. I understand why you might do this - showing overlaps with the same colour is impossible. A possible workaround could be to change the colour of the self-invocation bars, which was what I did. I will check with the teaching team and get back to you guys on this.

- Overall very well documented with UI screenshots. You might also want to include a class diagram.

**3.2. Poll feature**

- The first sequence diagram has a return arrow which is extended too long.

image

- There seems to be a typo in your alternatives "Alternative 1::". There are repeated colons. Also, it would be good to indicate your current choice of implementation e.g. Alternative 1(Current Choice).

- Overall well documented.  It would be good to include UI screenshots and/or a class diagram.

**3.3. Schedule feature**

- There seems to be a typo in your alternatives "Alternative 1::". There are repeated colons. Also, it would be good to indicate your current choice of implementation e.g. Alternative 1(Current Choice).

- Very detailed analysis on time and space complexity.

- Having some diagrams (class, sequence, or perhaps the "Architecture" of your schedule feature) might be useful to aid your explanations.

**3.4. TimeTable feature**

- Minor typos in your explanation: "All of these must be **fulfiled** or else a Invalid Nusmods link will be **returend** to the application."

- Might want to consider adding return arrows on self-invocation in sequence diagram.

- Would be good to have alternative implementations and evaluate their Pros and Cons.

**3.5. Add/Delete Friend feature**

- The text in the sequence diagram is really small in my opinion. As mentioned earlier, I can always zoom in, but not sure if you'll be penalised. 

- Try to render the diagram without all the red "spellcheck" underlines. 

- Might be good to add a bigger gap between Person p1 and p2 so that you can draw the lifeline. At this point I can't really tell whether they continue to live on or a destroyed. 

- **IMPORTANT:** In your add friend command, do you **create** 2 new Persons or do you **use existing** Persons? Your diagram seems to imply that you are creating 2 new Persons. 

- Having a class diagram might be helpful. Your guide does mention "we represents a friend of a Person using a Friend class." and it would be easier to see how it is "linked" to other classes.

- Overall well documented, but it would be good if you listed alternative implementations and their Pros and Cons.

**Appendix H: Instructions for Manual Testing**

- Don't forget to update this section.

**Overall Comments**

In general, your developer guide is on track. Minor amendments and inclusions can be added to make it better. Also, take note of formatting and ensure that it is consistent. For example, in Section 3, "Login/logout feature" does not bold the Alternatives, but other features do.

**Comment by AceCentury** [comment]
**Important**

Stacked activation bars for self-invocation **are wrong**. Example:
image

I would recommend using different colours for self-invocation activation bars so that you may show the overlap accordingly. An example is in the website/textbook:

image


I have also previously commented that some of your diagrams are very small, especially if the interactions spans across components. For those diagrams, please refer to this clarification I received from Professor

"Reference frames are an optional topic -- peer reviewers might not understand them. The better thing to do is not to show low-level interactions across components. instead, they should do multi-level diagrams. High-level diagram at component level, and further diagrams to show what happens inside each component, if necessary."

In short, while reference frames _can_ be used to solve the issue of having too much information in a diagram, it might cause some confusion for peer reviewers. You are advised to do multi-level diagrams. Take note of this, especially for your sequence diagrams.

**Comment by yaofeng-wang** [comment]
@AceCentury Thank you for the review! Will update them ASAP!

**Comment by yaofeng-wang** [comment]
@AceCentury 
For login/logout feature:
1. Updated both sequence diagrams.
2. Added manual tests.
3. Corrected inconsistencies in format for "Alternatives" part.

Did not include a class diagram because for this feature, I've merely add in additional methods/attributes into classes already existing in AddressBook Level 4 e.g. modelManager, Person, etc. If I were to include a class diagram, most of the methods/attributes would have existed in AddressBook Level 4. Would you still recommend including a class diagram?

**Comment by AceCentury** [comment]
Developer Guide review performed on Thursday (1 day after tutorial) as requested by team.

**Disclaimer**: This final review is meant to be brief check to spot any major errors or potential issues which need to be fixed. It will not be as comprehensive as my previous reviews where I point out other minor errors (such as formatting etc.) The onus is on you guys to thoroughly proofread and make the necessary amendments, if any.

**Poll Feature**

- It would be good if the text in the sequence diagram could be made bigger.

**TimeTable Feature**

- The sequence diagram is very small. Try to make it bigger (including the text). 

**Add/Delete Friend Feature**

- The text used for function calls and return values is extremely small. Try to make the text larger. I can't read it without zooming in.

**Comment by AceCentury** [comment]
> @AceCentury
> For login/logout feature:
> 
> 1. Updated both sequence diagrams.
> 2. Added manual tests.
> 3. Corrected inconsistencies in format for "Alternatives" part.
> 
> Did not include a class diagram because for this feature, I've merely add in additional methods/attributes into classes already existing in AddressBook Level 4 e.g. modelManager, Person, etc. If I were to include a class diagram, most of the methods/attributes would have existed in AddressBook Level 4. Would you still recommend including a class diagram?

@yao-feng Having a class diagram would help the markers gain a better understanding of your UML Modelling skills. Also, while your implementation _may_ involve classes already existing in AB4, does your feature _create_ more functionality by being associated to those classes? 

Something to think about: Do you see a class diagram for the ModelManager or Person class anywhere in the developer guide? Assuming I were a new developer on your team with **no background** on what is AB4, would I know how your feature makes associations between those classes just by reading your documentation?

From the website: https://nus-cs2103-ay1819s1.github.io/cs2103-website/admin/project-deliverables.html

> Contributions to the Developer Guide: Reproduce the parts in the Developer Guide that you wrote. Ensure there is enough content to evaluate your technical documentation skills and UML modelling skills. You can include descriptions of your design/implementations, possible alternatives, pros and cons of alternatives, etc.

**Comment by AceCentury** [comment]
Added v1.3 reviewed label as I forgot to add it the last time I reviewed >_<

**Comment by yaofeng-wang** [comment]
@AceCentury Got it. Thanks for the review :thumbsup:
